pandas/_libs/tslibs/period.pyx

```diff
@@ -806,24 +806,22 @@ cdef int64_t get_period_ordinal(npy_datetimestruct *dts, int freq) nogil:
         return unix_date_to_week(unix_date, freq - FR_WK)
 
 
-cdef void get_date_info(int64_t ordinal, int freq,
-                        npy_datetimestruct *dts) nogil:
+cdef void get_date_info(int64_t ordinal, int freq, npy_datetimestruct *dts) nogil:
     cdef:
-        int64_t unix_date
-        double abstime
+        int64_t unix_date, nanos
+        npy_datetimestruct dts2
 
     unix_date = get_unix_date(ordinal, freq)
-    abstime = get_abs_time(freq, unix_date, ordinal)
-
-    while abstime < 0:
-        abstime += 86400
-        unix_date -= 1
+    nanos = get_time_nanos(freq, unix_date, ordinal)
 
-    while abstime >= 86400:
-        abstime -= 86400
-        unix_date += 1
+    pandas_datetime_to_datetimestruct(unix_date, NPY_FR_D, dts)
 
-    date_info_from_days_and_time(dts, unix_date, abstime)
+    dt64_to_dtstruct(nanos, &dts2)
+    dts.hour = dts2.hour
```

Comment:
Any reason you don't just pass the reference to dts in the call above?

Reply:
Because the caller might still need `dts`

Reply:
Doesn't calling dt64_to_dtstruct do exactly what the next 5 lines of code do anyway?

Reply:
i can confirm that just passing `dts` does Break The World.  IIRC part of whats going on here is that `dts` may be out of bounds for dt64[ns]

Reply:
Ah OK. Thanks for checking - seems strange to have to do this but probably a separate issue then